consult-lsp.el

```diff
@@ -322,6 +336,97 @@ CURRENT-WORKSPACE? has the same meaning as in `lsp-diagnostics'."
      :narrow (consult--type-narrow consult-lsp--symbols--narrow)
      :state (consult-lsp--symbols--state))))
 
+
+;;;; File symbols
+
+(defun consult-lsp--flatten-document-symbols (to-flatten flattened)
+  "Helper function for flattening document symbols to a plain list."
+  (dolist (table to-flatten)
+    (when (hash-table-p table)
+      (push table flattened)
+      (when-let ((children (gethash "children" table)))
+        (setq flattened (consult-lsp--flatten-document-symbols
+                         (append children nil) ; convert children from vector to list
+                         flattened))
+        (remhash "children" table))))
+  flattened)
+
+(defun consult-lsp--file-symbols-candidates ()
+  "Returns all candidates for a `consult-lsp-file-symbols' search."
+  (consult--forbid-minibuffer)
+  (let ((all-symbols (consult-lsp--flatten-document-symbols
+                      (lsp-request "textDocument/documentSymbol"
+                                   (lsp-make-document-symbol-params :text-document
+                                                                    (lsp--text-document-identifier)))
+                      nil))
+        (candidates))
+    (dolist (symbol all-symbols)
+      (let ((line (thread-first symbol
+                    (lsp:document-symbol-selection-range)
+                    (lsp:range-start)
+                    (lsp:position-line)
+                    (lsp-translate-line)))
+            (cbeg (thread-first symbol
+                    (lsp:document-symbol-selection-range)
+                    (lsp:range-start)
+                    (lsp:position-character)
+                    (lsp-translate-column)))
+            (cend (thread-first symbol
+                    (lsp:document-symbol-selection-range)
+                    (lsp:range-end)
+                    (lsp:position-character)
+                    (lsp-translate-column))))
+        (let ((beg (lsp--line-character-to-point line cbeg))
+              (end (lsp--line-character-to-point line cend))
+              (marker (make-marker)))
+          (set-marker marker beg)
+          ;; Pre-condition to respect narrowing
+          (unless (or (< beg (point-min))
+                      (> end (point-max)))
+            (push (consult--location-candidate
+                   (consult--buffer-substring beg end 'fontify)
+                   marker
+                   (1+ line)
+                   'consult--type (lsp:symbol-information-kind symbol)
+                   'consult--name (lsp:symbol-information-name symbol))
+                  candidates)))))
+    (unless candidates
+      (user-error "No symbols"))
+    candidates))
+
+(defun consult-lsp--file-symbols-annotate ()
+  "Annotation function for `consult-lsp-file-symbols'."
+  (when consult-lsp-use-marginalia
+    (require 'marginalia))
```

Comment:
This check shouldn't be necessary, because a user who wants to use marginalia is going to activate the minor-mode anyway, thus automatically requiring `'marginalia`